pkgs/build-support/vm/default.nix

```diff
@@ -1,25 +1,72 @@
 { lib
+
+# pkgs used inside the VM
 , pkgs
+
+# pkgs used outside the VM
+# (similar to NixOS's `virtualisation.host.pkgs` option)
+# useful for easily running VMs on Darwin
+, hostPkgs ? pkgs.buildPackages
+
 , customQemu ? null
 , kernel ? pkgs.linux
 , img ? pkgs.stdenv.hostPlatform.linux-kernel.target
 , storeDir ? builtins.storeDir
-, rootModules ?
-    [ "virtio_pci" "virtio_mmio" "virtio_blk" "virtio_balloon" "virtio_rng" "ext4" "unix" "virtiofs" "crc32c_generic" ]
+, rootModules ? null
 }:
 
 let
   inherit (pkgs) bash bashInteractive busybox cpio coreutils e2fsprogs fetchurl kmod rpm
     stdenv util-linux
-    buildPackages writeScript writeText runCommand;
+    buildPackages writeScript runCommand;
+
+  hostSpecifics = if !hostPkgs.stdenv.isDarwin then
+    {
+      # Host is Linux
+      # - Use virtiofs as it is faster than 9p
+      defaultFilesystemRootModules = [ "virtiofs" ];
+      mountOptions = "-t virtiofs";
+      qemuOptions = [
+        "-chardev socket,id=store,path=virtio-store.sock"
+        "-device vhost-user-fs-pci,chardev=store,tag=store"
+        "-chardev socket,id=xchg,path=virtio-xchg.sock"
+        "-device vhost-user-fs-pci,chardev=xchg,tag=xchg"
+      ];
+      preQemuComamnds = ''
+        ${pkgs.virtiofsd}/bin/virtiofsd --xattr --socket-path virtio-store.sock --sandbox none --shared-dir "${storeDir}" &
+        ${pkgs.virtiofsd}/bin/virtiofsd --xattr --socket-path virtio-xchg.sock --sandbox none --shared-dir xchg &
+      '';
+      additionalQEMU_OPTS = { memSize }: "-object memory-backend-memfd,id=mem,size=${toString memSize}M,share=on -machine memory-backend=mem";
+    }
+  else
+    {
+      # Host is Darwin
+      # - Use 9p as virtiofsd isn't supported on Darwin
+      defaultFilesystemRootModules = [ "9p" "9pnet_virtio" ];
+      mountOptions = "-t 9p -o trans=virtio,version=9p2000.L,cache=loose,msize=131072";
+      qemuOptions = [
+        "-virtfs local,path=${storeDir},security_model=none,mount_tag=store"
+        "-virtfs local,path=xchg,security_model=none,mount_tag=xchg"
+      ];
+      preQemuComamnds = "";
+      additionalQEMU_OPTS = { memSize }: "";
+    };
 in
 rec {
   qemu-common = import ../../../nixos/lib/qemu-common.nix { inherit lib pkgs; };
 
-  qemu = buildPackages.qemu_kvm;
+  qemu = hostPkgs.qemu_kvm;
+
+  # export hostPkgs so that downstream tools (e.g. disko) that wrap around e.g. runInLinuxVM can use it
+  inherit hostPkgs;
 
   modulesClosure = pkgs.makeModulesClosure {
-    inherit kernel rootModules;
+    inherit kernel;
+    rootModules = if rootModules == null then
+      [ "virtio_pci" "virtio_mmio" "virtio_blk" "virtio_balloon" "virtio_rng" "ext4" "unix" "crc32c_generic" ]
+        ++ hostSpecifics.defaultFilesystemRootModules
+    else
+      rootModules;
     firmware = kernel;
```

Comment:
@Mic92 I see this logic is duplicated in [disko](https://github.com/nix-community/disko/blob/a08bfe06b39e94eec98dd089a2c1b18af01fef19/lib/make-disk-image.nix#L17) as well. Can we do something about that? 

The "root" of having this duplication is that disko definitely needs to be able to set `rootModules`, right? Maybe `runInLinuxVM` should always add the `rootModules` it needs for sharing the store?

Reply:
@felixscheinost Would be cool if we could only add to the list, rather having to overwrite it, yes.